### PR TITLE
implemented MyLinkedList;

### PR DESCRIPTION
**Pull request by aksoli666**
override methods: add(T value),
add(T value, int index),
addAll(), get(), set(),
remove(int index),
remove(T object), size(), isEmpty();
created methods: getNodeAtIndex(),
checkIndexInclusive/Exclusive(),
removeFirst/Last(), unlink()
created class: Node<T>

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,45 +3,204 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private Node<T> newNode;
+    private Node<T> currentNode;
+    private Node<T> removed;
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/MyLinkedList.java:10]: Why you save it in field?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,45 +3,204 @@
...
+
+        private Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.item = value;
+            this.next = next;
+        }
+
+        private Node(T value) {
+            this.item = value;
+        }
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/MyLinkedList.java:204]: You don`t need it

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,45 +3,204 @@
...
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/MyLinkedList.java:43]: You have same logic like in add(value) method, you can combine logic and create private method

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,45 +3,204 @@
...
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInclusive(index);
+
+        currentNode = getNodeAtIndex(index);
+
+        return currentNode.item;
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/MyLinkedList.java:68]: For what empty lines?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,45 +3,204 @@
...
     }
 
     @Override
     public boolean remove(T object) {
+        currentNode = head;
+
+        while (currentNode != null) {
+            if ((currentNode.item == null && object == null)
```
Comment (Sanderfuck) [inline src/main/java/core/basesyntax/MyLinkedList.java:106]: Check currentNode.item reference and object

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,45 +3,197 @@
...
+        }
+
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexExclusive(index);
+
+        Node<T> newNode = new Node<>(head, value, tail);
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:29]: why head and tail?

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 7]
```diff
@@ -3,45 +3,197 @@
...
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInclusive(index);
+
+        Node<T> currentNode = getNodeAtIndex(index);
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:63]: This is code repeating. You always check the index before getting node by index. What you should do is move the check logic INSIDE the get method.

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 8]
```diff
@@ -3,45 +3,197 @@
...
+        return removeNode(currentNode) ? currentNode.item : null;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+
+        while (currentNode != null) {
+            if ((currentNode.item == object)
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:99]:
```suggestion
            if (currentNode.item == object
```

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 9]
```diff
@@ -3,45 +3,197 @@
...
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeAtIndex(int index) {
+        Node<T> currentNode = head;
+
+        for (int i = 0; i < index; i++) {
+            currentNode = currentNode.next;
+        }
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:123]: image

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 10]
```diff
@@ -3,45 +3,197 @@
...
+        } else if (currentNode == tail) {
+            return removeLast() != null;
+        } else {
+            unlink(currentNode);
+            size--;
+            return true;
+        }
+    }
+
+    private void unlink(Node<T> currentNode) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:183]: delete this method

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 11]
```diff
@@ -3,45 +3,197 @@
...
+        if (index == 0) {
+            return removeFirst();
+        }
+
+        if (index == size - 1) {
+            return removeLast();
+        }
+
+        Node<T> currentNode = getNodeAtIndex(index);
+        return removeNode(currentNode) ? currentNode.item : null;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:91]:
```suggestion
        Node<T> currentNode = getNodeAtIndex(index);
        removeNode(currentNode);
        return  currentNode.item;
```

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by kerrrusha — APPROVED** [review]

**Pull request closed without merging**